### ISSUE TITLE
TypeScript generates invalid `import` for re-exported types

### ISSUE THREAD

**Issue by raymondfeng** [Microsoft/TypeScript#26985]
The issue was reported on https://github.com/Microsoft/TypeScript/pull/24941#issuecomment-415900977. It was initially discovered in strongloop/loopback-next#1643.

In this case, MetadataAccessor is exported from @loopback/metadata and re-exported by @loopback/context. Please note that @loopback/repository-json-schema module imports MetadataAccessor from @loopback/context. It leads to incorrectly generated code.

**TypeScript Version:**  3.0.x and up

**Search Terms:** TypeScript import types

**Code**

https://github.com/strongloop/loopback-next/blob/5b27323ba420bfc500fdf9f2762e7292b96ffa0f/packages/repository-json-schema/src/keys.ts

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
```

**Expected behavior:**

TypeScript generates the following code:
```ts
import { MetadataAccessor } from '@loopback/context';
import { JSONSchema6 as JSONSchema } from 'json-schema';
/**
 * Metadata key used to set or retrieve repository JSON Schema
 */
export declare const JSON_SCHEMA_KEY: MetadataAccessor<JSONSchema, import("@loopback/context").DecoratorType>;
```
**Actual behavior:**

TypeScript generates invalid code as follows:
```ts
import { MetadataAccessor } from '@loopback/context';
import { JSONSchema6 as JSONSchema } from 'json-schema';
/**
 * Metadata key used to set or retrieve repository JSON Schema
 */
export declare const JSON_SCHEMA_KEY: MetadataAccessor<JSONSchema, import("../context/node_modules/@loopback/metadata/src/types").DecoratorType>;
```
**Playground Link:** 

**Related Issues:** 
https://github.com/Microsoft/TypeScript/pull/24874

**Comment by raymondfeng** [comment]
@weswigham I created a simpler project to reproduce the problem - https://github.com/raymondfeng/ts-26985.

**Comment by mbrowne** [comment]
Is it expected behavior that the PR that was just merged for this (#27340) seems to break certain `export const` statements in a project with `"declaration": true` set in tsconfig?

For example:
```
error TS2742: The inferred type of 'Wrapper' cannot be named without a reference to '../../../../../../../../../../Users/mbrowne/GoogleDrive/www/artnet-styled/styled/node_modules/@types/styled-components'. This is likely not portable. A type annotation is necessary.

13 export const Wrapper = styled.div`
```

**Comment by weswigham** [comment]
Yup. When that error occurs, we would otherwise have written a reference to the module specifier in the message (and you see how bad that is, right?). It's better to require an annotation when we detect a required path like that.
